builtin/providers/vcd/resource_vcd_network.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		fmt.Printf("error: %v\n", err)
+	}
+
+	newnetwork := &types.OrgVDCNetwork{
+		Xmlns: "http://www.vmware.com/vcloud/v1.5",
+		Name:  d.Get("name").(string),
+		Configuration: &types.NetworkConfiguration{
+			FenceMode: d.Get("fence_mode").(string),
+			IPScopes: &types.IPScopes{
+				IPScope: types.IPScope{
+					IsInherited: false,
+					Gateway:     d.Get("gateway").(string),
+					Netmask:     d.Get("netmask").(string),
+					DNS1:        d.Get("dns1").(string),
+					DNS2:        d.Get("dns2").(string),
+					DNSSuffix:   d.Get("dns_suffix").(string),
+					IPRanges:    &ipRanges,
+				},
+			},
+			BackwardCompatibilityMode: true,
+		},
+		EdgeGateway: &types.Reference{
+			HREF: edgeGateway.EdgeGateway.HREF,
+		},
+		IsShared: false,
+	}
+
+	log.Printf("[INFO] NETWORK: %#v", newnetwork)
+
+	err = retryCall(4, func() error {
+		return vcd_client.OrgVdc.CreateOrgVDCNetwork(newnetwork)
+	})
+	if err != nil {
+		return fmt.Errorf("Error: %#v", err)
+	}
+
+	err = vcd_client.OrgVdc.Refresh()
+	if err != nil {
+		return fmt.Errorf("Error refreshing vdc: %#v", err)
+	}
+
+	network, err := vcd_client.OrgVdc.FindVDCNetwork(d.Get("name").(string))
+	if err != nil {
+		return fmt.Errorf("Error finding network: %#v", err)
+	}
+
+	if dhcp, ok := d.GetOk("dhcp_pool"); ok {
+		err = retryCall(4, func() error {
+			task, err := edgeGateway.AddDhcpPool(network.OrgVDCNetwork, dhcp.(*schema.Set).List())
+			if err != nil {
+				return fmt.Errorf("Error adding DHCP pool: %#v", err)
+			}
+
+			return task.WaitTaskCompletion()
+		})
+		if err != nil {
+			return fmt.Errorf("Error completing tasks: %#v", err)
+		}
+
+	}
+
+	d.SetId(d.Get("name").(string))
+
+	return resourceVcdNetworkRead(d, meta)
+}
+
+func resourceVcdNetworkUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	vcd_client := meta.(*govcd.VCDClient)
+
+	log.Printf("[DEBUG] VCD Client configuration: %#v", vcd_client)
+	return nil
+}
+
+func resourceVcdNetworkRead(d *schema.ResourceData, meta interface{}) error {
+	vcd_client := meta.(*govcd.VCDClient)
+	log.Printf("[DEBUG] VCD Client configuration: %#v", vcd_client)
+	log.Printf("[DEBUG] VCD Client configuration: %#v", vcd_client.OrgVdc)
+
+	err := vcd_client.OrgVdc.Refresh()
+	if err != nil {
+		return fmt.Errorf("Error refreshing vdc: %#v", err)
+	}
+
+	network, err := vcd_client.OrgVdc.FindVDCNetwork(d.Id())
+	if err != nil {
+		log.Printf("[DEBUG] Network no longer exists. Removing from tfstate")
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("name", network.OrgVDCNetwork.Name)
+	d.Set("href", network.OrgVDCNetwork.HREF)
+	d.Set("fence_mode", network.OrgVDCNetwork.Configuration.FenceMode)
+	d.Set("gateway", network.OrgVDCNetwork.Configuration.IPScopes.IPScope.Gateway)
+	d.Set("netmask", network.OrgVDCNetwork.Configuration.IPScopes.IPScope.Netmask)
+	d.Set("dns1", network.OrgVDCNetwork.Configuration.IPScopes.IPScope.DNS1)
+	d.Set("dns2", network.OrgVDCNetwork.Configuration.IPScopes.IPScope.DNS2)
```

Comment:
Does the client library guarantee that these deeply nested structures are always populated? If not we probably need to add more `nil` checking here to avoid panics.